File: app/services/customer_service.py
from typing import Optional

from sqlalchemy.orm import Session

from app.models.customer import Customer, CustomerType
from app.repositories.customer_repo import CustomerRepository
from app.schemas.customer import CustomerCreateRequest, CustomerUpdateRequest
from app.core.custom_response import CustomResponse
from app.core.http_constants import HttpConstants

C = HttpConstants.HttpResponseCodes


class CustomerService:
    def __init__(self, db: Session):
        self.repo = CustomerRepository(db)

    # ── Helpers ───────────────────────────────────────────────────────

    def _generate_unique_customer_id(self) -> str:
        for _ in range(10):
            candidate = Customer.generate_customer_id()
            if not self.repo.customer_id_exists(candidate):
                return candidate
        raise RuntimeError("Could not generate a unique customer ID after 10 attempts.")

    def _classify_customer_type(self, phone: str) -> CustomerType:
        existing = self.repo.get_by_phone(phone)
        return CustomerType.regular if existing else CustomerType.new

    # ── CRUD ──────────────────────────────────────────────────────────

    def create_customer(self, payload: CustomerCreateRequest) -> CustomResponse:
        if self.repo.get_by_phone(payload.phone):
            return CustomResponse(C.CONFLICT, f"A customer with phone '{payload.phone}' already exists.")

        if payload.email and self.repo.get_by_email(payload.email):
            return CustomResponse(C.CONFLICT, f"A customer with email '{payload.email}' already exists.")

        customer = Customer(
            customer_id   = self._generate_unique_customer_id(),
            name          = payload.name,
            phone         = payload.phone,
            email         = payload.email,
            address       = payload.address,
            date_of_birth = payload.date_of_birth,
            notes         = payload.notes,
           customer_type=payload.customer_type,
        )
        customer = self.repo.create(customer)
        return CustomResponse(C.CREATED, "Customer created successfully", data=customer)

    def get_customer_by_id(self, id: int) -> CustomResponse:
        customer = self.repo.get_by_id(id)
        if not customer:
            return CustomResponse(C.NOT_FOUND, f"Customer with id {id} not found.")
        return CustomResponse(C.OK, "Customer fetched successfully", data=customer)

    def get_all_customers(
        self,
        page: int = 1,
        limit: int = 10,
        customer_type: Optional[str] = None,
        search: Optional[str] = None,
    ) -> CustomResponse:
        records, total = self.repo.get_all(page=page, limit=limit, customer_type=customer_type, search=search)
        return CustomResponse(
            C.OK,
            "Customers fetched successfully",
            data=records,
            meta={
                "total":       total,
                "page":        page,
                "limit":       limit,
                "total_pages": (total + limit - 1) // limit,
                "has_next":    page * limit < total,
                "has_prev":    page > 1,
            },
        )

    def update_customer(self, id: int, payload: CustomerUpdateRequest) -> CustomResponse:
        customer = self.repo.get_by_id(id)
        if not customer:
            return CustomResponse(C.NOT_FOUND, f"Customer with id {id} not found.")

        if payload.email and payload.email != customer.email:
            if self.repo.get_by_email(payload.email):
                return CustomResponse(C.CONFLICT, f"Email '{payload.email}' is already in use.")

        for field, value in payload.model_dump(exclude_unset=True).items():
            setattr(customer, field, value)

        customer = self.repo.update(customer)
        return CustomResponse(C.OK, "Customer updated successfully", data=customer)

    def toggle_status(self, id: int) -> CustomResponse:
        customer = self.repo.get_by_id_any_status(id)
        if not customer:
            return CustomResponse(C.NOT_FOUND, f"Customer with id {id} not found.")
        customer = self.repo.set_active_status(customer, not customer.is_active)
        status_msg = "activated" if customer.is_active else "deactivated"
        return CustomResponse(C.OK, f"Customer {status_msg} successfully", data=customer)

    def search_by_phone(self, phone: str) -> CustomResponse:
        customer = self.repo.get_by_phone(phone)
        if not customer or not customer.is_active:
            return CustomResponse(C.NOT_FOUND, f"No active customer found with phone '{phone}'.")
        return CustomResponse(C.OK, "Customer fetched successfully", data=customer)
    
    def delete_customer(self, customer_id: int) -> CustomResponse:
      customer = self.repo.get_by_id(customer_id)
      if not customer:
        return CustomResponse(C.NOT_FOUND, "Customer not found")
      self.repo.delete(customer)
      return CustomResponse(C.OK, "Customer deleted successfully")

File: src/app/core/services/customer.service.ts
import { Injectable } from '@angular/core';
import { HttpResponse } from '@angular/common/http';
import { GenericService } from './generic.service';
import { Observable } from 'rxjs';

/** Request body for POST /customers */
export interface CreateCustomerPayload {
  name: string;
  phone: string;
  email: string;
  address: string;
  date_of_birth: string;
  notes: string;
  customer_type: 'new' | 'regular' | 'vip';
}

export interface UpdateCustomerPayload {
  name?: string;
  phone?: string;
  email?: string;
  address?: string;
  date_of_birth?: string;
  notes?: string;
  customer_type?: 'new' | 'regular' | 'vip';
}

export interface CustomerSearchPayload {
  page: number;
  page_size: number;
  search?: string;
}

export interface CustomerApiItem {
  customer_id: string;
  phone: string;
  id: number;
  address: string;
  notes: string;
  is_active: boolean;
  updated_at: string;
  name: string;
  email: string;
  date_of_birth: string;
  customer_type: string;
  registered_at: string;
}

export interface CustomerPaginationResponse {
  success: boolean;
  statusCode: number;
  message: string;
  data: CustomerApiItem[];
  meta: {
    total: number;
    page: number;
    limit: number;
    total_pages: number;
    has_next: boolean;
    has_prev: boolean;
  };
}

export interface CustomerSearchResponse {
  success?: boolean;
  statusCode?: number;
  message?: string;
  data: CustomerApiItem[];
}

@Injectable({
  providedIn: 'root'
})
export class CustomerService {

  constructor(
    private genricService: GenericService,
  ) {}

  createCustomer(payload: CreateCustomerPayload): Observable<HttpResponse<unknown>> {
    return this.genricService.Post('customers', payload, { observe: 'response' });
  }

  updateCustomer(id: number, payload: UpdateCustomerPayload): Observable<HttpResponse<unknown>> {
    return this.genricService.Put(`customers/${id}`, payload);
  }

  /** DELETE /customers/{id} */
  deleteCustomer(id: number): Observable<unknown> {
    return this.genricService.DeleteRequest(`customers/${id}`);
  }

  customerPagination(payload: CustomerSearchPayload): Observable<CustomerPaginationResponse> {
    const search = payload.search?.trim() ?? '';
    const query = search
      ? `customers?page=${payload.page}&page_size=${payload.page_size}&search=${encodeURIComponent(search)}`
      : `customers?page=${payload.page}&page_size=${payload.page_size}`;
    return this.genricService.Get(query);
  }

  getAllCustomers(): Observable<CustomerSearchResponse> {
    return this.genricService.Get<CustomerSearchResponse>('customers/search');
  }

}
